File: 532. K-diff Pairs in an Array.py
class Solution(object):
    def findPairs(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        cnt = 0
        if k == 0:
            num, cnt = set(nums), 0
            lens, lenn = len(nums), len(num)
            if lenn != lens:  # there are duplicate in the array
                for i in num:
                    if nums.count(i) > 1:
                        cnt += 1
        elif k > 0:
            num = set(nums)
            lenn, num, cnt = len(num), list(num), 0
            num.sort()
            for i in xrange(lenn - 1): # use two pointers to count
                for j in xrange(i + 1, lenn):
                    if num[j] - num[i] == k:
                        cnt += 1
                        break
                    elif num[j] - num[i] > k:
                        break
        return cnt # k < 0 

File: 532. K-diff Pairs in an Array.cs
public class Solution {
    public int FindPairs(int[] nums, int k) {
        if (nums.Length == 0) {
            return 0;
        }
        Array.Sort(nums);
        int cnt = 0, tmp = nums[0] - 1;
        if (k == 0) {
            for (int i = 0; i < nums.Length - 1; tmp = nums[i], i++) {
                if (nums[i] == tmp) {
                    continue;
                }
                if (nums[i + 1] == nums[i]) {
                    cnt++;
                }
            }
        }
        if (k > 0) {
            for (int i = 0; i < nums.Length - 1; tmp = nums[i], i++) {
                if (tmp == nums[i]) {
                    continue;
                }
                for (int j = i + 1; j < nums.Length; j++) {
                    if (nums[j] - nums[i] < k) {
                        continue;
                    }
                    if (nums[j] - nums[i] == k) {
                        cnt++;
                        break;
                    }
                    if (nums[j] - nums[i] > k) {
                        break;
                    }
                }
            }
        }
        return cnt;
    }
}
